serverlessworkflow/modules/ROOT/pages/cloud/kubernetes-service-discovery.adoc

```diff
@@ -0,0 +1,212 @@
+====
+The **Kubernetes GVK** (Group, Version and Kind) and the **Name** values are **mandatory** and can't be empty, otherwise, the engine will throw a validation issue and the discovery will not be performed.  +
+The **Namespace** is **optional**. In case of an empty value, the current namespace or context will be used.
+
+For **OpenShift** specific resources, **Deployment Config** or **Routes**, you will be required to add the following dependency in your Application's classpath:
+
+[source,xml,subs="attributes+"]
+----
+<dependency>
+  <groupId>io.quarkus</groupId>
+  <artifactId>quarkus-openshift-client</artifactId>
+</dependency>
+----
+
+It can be added with link:{quarkus_cli_url}[Quarkus CLI]:
+[source,shell]
+----
+quarkus extension add quarkus-openshift-client
+----
+====
+
+=== Additional Query Parameters
+
+Also known as Query String, are defined in the very same way that we are already used to do with URLs to assign value
+to specific attributes.
+
+There are available two query parameters to help the engine to be more precise when querying for the given Kubernetes Resource,
+which are:
+
+* Custom Labels: Can be used to filter services in case there are more than one Service with the same `Label Selector`
+but exposing different ports. In this case we can instruct the engine that, when more than 1 service is found, use the one
+that have the provided label, the label is defined with the following expression, for multiple labels use `semicolon`:
+** `labels=label-name=namevalue;another-label=another-value`
+** Example:
++
+[source,shell]
+----
+kubernetes:v1/pod/<namespace>/<pod-name>?labels=label-name=test-label
+----
+*** With the URI above we will instruct that, if there are more than 1 service exposing the given pod, the label
+`label-name=test-label` will be used to filter the service. If the label does not exist, the first found service will be used.
+
+* Custom port name: Useful for determine which port to use when there are many ports configured in the target Service or Container.
+It can be configured using the following patter:
+** `port-name=<PORT_NAME>`: the port name to be queried.
+
+
+== Configuration
+
+There are no specific configuration for the Discovery to work properly except by using the expected URI pattern.
+However, the okhttp communication interceptor, which logs the communication between the application and the Kubernetes API +
+is disabled by default, if there is a need to debug it, it can be enabled back just by setting this application property:
+
+[source,shell]
+----
+quarkus.log.category."okhttp3.OkHttpClient".level=INFO
+----
+
+
+== How does it work?
+
+During application startup, the service discovery is performed at the Quarkus _STATIC_INIT_ time. +
+First, it scans the Quarkus configuration values and searches for the Kubernetes URI pattern. +
+Then if at least one is found, the engine will parse the URI, query the Kubernetes API for the given resource, and override the given property.
```

Comment:
it the KubeURI is not found the application flow will continue as usual.

Reply:
ok. I think we should add a `fail-fast` possibility in the future ? because if the application cannot retrieve the URI, then it may not be able to work properly after a start, as this feature is only done at startup. wdyt ?

Reply:
if the KubeURI is not found, we do expect user to use the dynamic http url for the service.
In the other hand, we could add the fail-fast  feature if user sets the KubeURI and for any reason it fails to be discovered, parsed, etc.
@fjtirado and I have discussed about it and we thought better just log the error and don't stop the app startup.

Reply:
oki then I will create an enhancement with a fail-fast option

Reply:
to fail fast it, we just need to return null in the Exception catch in the properties interceptor. We can make it configurable though.

Reply:
@spolti Created https://issues.redhat.com/browse/KOGITO-7915 and assigned it to you